contracts/ReputationMiningCycle.sol

```diff
@@ -143,56 +240,488 @@ contract ReputationMiningCycle {
       }
 
       // Note that two hashes have completed this challenge round (either one accepted for now and one rejected, or two rejected)
       nHashesCompletedChallengeRound[round] += 2;
 
       // Punish the people who proposed the hash that was rejected
-      IColonyNetwork(colonyNetworkAddress).punishStakers(submittedHashes[disputeRounds[round][idx].hash][disputeRounds[round][idx].nNodes]);
-
+      IColonyNetwork(colonyNetworkAddress).punishStakers(submittedHashes[disputeRounds[round][idx].proposedNewRootHash][disputeRounds[round][idx].nNodes]);
     }
     //TODO: Can we do some deleting to make calling this as cheap as possible for people?
   }
+
+  /// @notice Respond to a binary search step, to eventually discover where two submitted hashes differ in their Justification trees.
+  /// @param round The round number the hash we are responding on behalf of is in
+  /// @param idx The index in the round that the hash we are responding on behalf of is in
+  /// @param jhIntermediateValue The contents of the Justification Tree at the key given by `targetNode` (see function description). The value of `targetNode` is computed locally to establish what to submit to this function.
+  /// @param branchMask The branchMask of the Merkle proof that `jhIntermediateValue` is the value at key `targetNode`
+  /// @param siblings The siblings of the Merkle proof that `jhIntermediateValue` is the value at key `targetNode`
+  function respondToBinarySearchForChallenge(uint256 round, uint256 idx, bytes jhIntermediateValue, uint branchMask, bytes32[] siblings) public {
+    // TODO: Check this challenge is active.
+    // This require is necessary, but not a sufficient check (need to check we have an opponent, at least).
+    require(disputeRounds[round][idx].lowerBound!=disputeRounds[round][idx].upperBound);
+
+    uint256 targetNode = add(disputeRounds[round][idx].lowerBound, sub(disputeRounds[round][idx].upperBound, disputeRounds[round][idx].lowerBound) / 2);
+    bytes32 jrh = disputeRounds[round][idx].jrh;
+
+    bytes memory targetNodeBytes = new bytes(32);
+    assembly {
+      mstore(add(targetNodeBytes, 0x20), targetNode)
+    }
+
+    verifyProof(jrh, targetNodeBytes, jhIntermediateValue, branchMask, siblings);
+    // If verifyProof hasn't thrown, proof is correct.
+    // Process the consequences
+    processBinaryChallengeSearchResponse(round, idx, jhIntermediateValue, targetNode);
+  }
+
+  uint constant __ROUND__ = 0;
+  uint constant __IDX__ = 1;
+  uint constant __REPUTATION_BRANCH_MASK__ = 2;
+  uint constant __AGREE_STATE_NNODES__ = 3;
+  uint constant __AGREE_STATE_BRANCH_MASK__ = 4;
+  uint constant __DISAGREE_STATE_NNODES__ = 5;
+  uint constant __DISAGREE_STATE_BRANCH_MASK__ = 6;
+  uint constant __PREVIOUS_NEW_REPUTATION_BRANCH_MASK__ = 7;
+  uint constant __REQUIRE_REPUTATION_CHECK__ = 8;
```

Comment:
`solium` doesn't like these constant names and rightly so, we can use something like a `U_` prefix maybe?  `U_ROUND`, `U_IDX` etc